src/System.Private.ServiceModel/src/System/ServiceModel/Channels/BufferedStream.cs

```diff
@@ -0,0 +1,1098 @@
+    /// The assumption here is you will almost always be doing a series of reads or writes, but rarely
+    /// alternate between the two of them on the same stream.
+    ///
+    /// Class Invariants:
+    /// The class has one buffer, shared for reading & writing.
+    /// It can only be used for one or the other at any point in time - not both.
+    /// The following should be true:
+    /// <![CDATA[
+    ///   * 0 <= _readPos <= _readLen < _bufferSize
+    ///   * 0 <= _writePos < _bufferSize
+    ///   * _readPos == _readLen && _readPos > 0 implies the read buffer is valid, but we're at the end of the buffer.
+    ///   * _readPos == _readLen == 0 means the read buffer contains garbage.
+    ///   * Either _writePos can be greater than 0, or _readLen & _readPos can be greater than zero,
+    ///     but neither can be greater than zero at the same time.
+    ///  ]]>
+    /// This class will never cache more bytes than the max specified buffer size.
+    /// However, it may use a temporary buffer of up to twice the size in order to combine several IO operations on
+    /// the underlying stream into a single operation. This is because we assume that memory copies are significantly
+    /// faster than IO operations on the underlying stream (if this was not true, using buffering is never appropriate).
+    /// The max size of this "shadow" buffer is limited as to not allocate it on the LOH.
+    /// Shadowing is always transient. Even when using this technique, this class still guarantees that the number of
+    /// bytes cached (not yet written to the target stream or not yet consumed by the user) is never larger than the 
+    /// actual specified buffer size.
+    /// </summary>
+    internal sealed class BufferedStream : Stream
+    {
+        private const int DefaultBufferSize = 4096;
+
+
+        private Stream _stream; // Underlying stream.  Close sets _stream to null.
+
+        private byte[] _buffer; // Shared read/write buffer.  Alloc on first use.
+
+        private readonly int _bufferSize; // Length of internal buffer (not counting the shadow buffer).
+
+        private int _readPos; // Read pointer within shared buffer.
+        private int _readLen; // Number of bytes read in buffer from _stream.
+        private int _writePos; // Write pointer within shared buffer.
+
+        private Task<int> _lastSyncCompletedReadTask; // The last successful Task returned from ReadAsync
+        // (perf optimization for successive reads of the same size)
+
+        private readonly SemaphoreSlim _sem = new SemaphoreSlim(1, 1);
+
+        public BufferedStream(Stream stream) : this(stream, DefaultBufferSize) {}
+
+
+        public BufferedStream(Stream stream, int bufferSize)
+        {
+            if (stream == null)
+                throw new ArgumentNullException("stream");
+
+            if (bufferSize <= 0)
+                throw new ArgumentOutOfRangeException("bufferSize");
+
+            Contract.EndContractBlock();
```

Comment:
BTW, one rule I've seen followed is to *not* throw ArgumentExceptions from internal code but only to Assert (unless the internal code is somehow callable through public code, like a virtual).  Otherwise you have dead code that will never execute.